TELCODOCS-2644 clarify the NUMA aware scheduler scoring behavior (#106102)
openshift/openshift-docs: modules/cnf-numa-resource-scheduling-strategies.adoc

:_mod-docs-content-type: CONCEPT
[id="cnf-numa-resource-scheduling-strategies_{context}"]
= NUMA resource scheduling strategies

When scheduling high-performance workloads, the secondary scheduler employs specific strategies to determine which worker node is best suited for the task based on its internal NUMA resource distribution. While the scheduler uses NUMA-level data to score and select a worker node, the actual resource pinning within that node is managed by the local Topology Manager and CPU Manager.

When a high-performance workload is scheduled in a NUMA-aware cluster, the following steps occur:

. *Node Selection*: The scheduler first selects a suitable worker node based on cluster-wide criteria. For example taints, labels, or resource availability.

. *NUMA-Aware Scoring*: After a worker node is selected, the scheduler evaluates the available resources within each worker node's NUMA zones. It applies a scoring strategy to select the worker node that best fits the desired resource distribution.

> . *Local Allocation*: Once the pod is assigned to a worker node, the node-level components (Topology Manager) perform the authoritative allocation of specific CPUs and memory. The scheduler does not influence this final selection.

Reviewer: Maybe

```suggestion
. *Local Allocation*: Once the pod is assigned to a worker node, the node-level components (CPU and Topology Managers) perform the authoritative allocation of specific CPUs and memory. The scheduler does not influence this final selection.
```

CPU manager picks the CPUs based on input from all the others like Topology manager, Memory manager (for hugepages affinity) and other device managers (eg. for SR-IOV affinity).

:_mod-docs-content-type: CONCEPT
[id="cnf-numa-resource-scheduling-strategies_{context}"]
= NUMA resource scheduling strategies

[role="_abstract"]
The secondary scheduler optimizes the placement of high-performance workloads by using NUMA-aware scoring strategies to select the most suitable worker nodes. This process assigns workloads to nodes with sufficient resources while allowing local node managers to handle final resource pinning.

When scheduling high-performance workloads, the secondary scheduler employs specific strategies to determine which worker node is best suited for the task based on its internal NUMA resource distribution. While the scheduler uses NUMA-level data to score and select a worker node, the actual resource pinning within that node is managed by the local Topology Manager and CPU Manager.

When a high-performance workload is scheduled in a NUMA-aware cluster, the following steps occur:

> . *Node filtering*: The scheduler first filters the entire cluster to find a shortlist of _feasible_ nodes. A node is only kept if it meets all requirements, such as matching labels, respecting taints and tolerations, and importantly having sufficient available resources within its specific NUMA zones. If a node cannot satisfy the workload's NUMA affinity, it is filtered out at this stage.

Reviewer 2: ```suggestion
. *Node filtering*: The scheduler first filters the entire cluster to find a shortlist of _feasible_ nodes. A node is only kept if it meets all requirements, such as matching labels, respecting taints and tolerations, and, importantly, having sufficient available resources within its specific NUMA zones. If a node cannot satisfy the workload's NUMA affinity, it is filtered out at this stage.
```

[cols="2,3,3", options="header"]
|===
|Strategy |Description |Outcome
|`LeastAllocated` |Favors worker nodes that contain NUMA zones with the most available resources. |Distributes workloads across the cluster to nodes with the highest available headroom.

> |`MostAllocated` |Favors worker nodes where the requested resources fit into NUMA zones that are already highly utilized.|Consolidates workloads on already utilized nodes, potentially leaving other nodes idle.

Reviewer 2: ```suggestion
|`MostAllocated` |Favors worker nodes that are already highly utilized.|Consolidates workloads on already utilized nodes, potentially leaving other nodes idle.
```
Wording seems a bit convoluted. Does this make it tidier?
